src/parser.rs

```diff
@@ -39,12 +43,84 @@ enum HtmlParserOption {
 pub enum XmlParseError {
   ///Parsing returned a null pointer as document pointer
   GotNullPointer,
+  ///Could not open file error.
+  FileOpenError,
+  ///Document too large for libxml2.
+  DocumentTooLarge,
 }
 
 impl fmt::Debug for XmlParseError {
   fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
     match *self {
       XmlParseError::GotNullPointer => write!(f, "Got a Null pointer"),
+      XmlParseError::FileOpenError => write!(f, "Unable to open path to file."),
+      XmlParseError::DocumentTooLarge => write!(f, "Document too large for i32."),
+    }
+  }
+}
+
+/// Default encoding when not provided.
+const DEFAULT_ENCODING: *const ::std::os::raw::c_char = ptr::null();
+
+/// Default URL when not provided.
+const DEFAULT_URL: *const ::std::os::raw::c_char = ptr::null();
+
+/// Open file function.
+fn xml_open(filename: &str) -> io::Result<*mut ::std::os::raw::c_void> {
+  let ptr = Box::into_raw(Box::new(fs::File::open(filename)?));
+  Ok(ptr as *mut ::std::os::raw::c_void)
+}
+
+/// Read callback for an FS file.
+unsafe extern "C" fn xml_read(
+    context: *mut ::std::os::raw::c_void,
+    buffer: *mut ::std::os::raw::c_char,
+    len: ::std::os::raw::c_int,
+) -> ::std::os::raw::c_int
+{
+  // Len is always positive, typically 40-4000 bytes.
+  let file = context as *mut fs::File;
+  let buf = slice::from_raw_parts_mut(buffer as *mut u8, len as usize);
+  match io::Read::read(&mut *file, buf) {
+    Ok(v) => v as ::std::os::raw::c_int,
+    Err(_) => -1,
+  }
+}
+
+type XmlReadCallback = unsafe extern "C" fn(
+  *mut ::std::ffi::c_void,
+  *mut ::std::os::raw::c_char,
+  ::std::os::raw::c_int
+) -> ::std::os::raw::c_int;
+
+/// Close callback for an FS file.
+unsafe extern "C" fn xml_close(
+    context: *mut ::std::os::raw::c_void,
+) -> ::std::os::raw::c_int
+{
+  // Take rust ownership of the context and then drop it.
+  let file = context as *mut fs::File;
+  let _ = Box::from_raw(file);
+  0
+}
+
+type XmlCloseCallback = unsafe extern "C" fn(
+  *mut ::std::ffi::c_void,
+) -> ::std::os::raw::c_int;
+
+
+///Convert usize to i32 safely.
+fn usize_to_i32(value: usize) -> Result<i32, XmlParseError> {
```

Comment:
Cool handler, just reread and saw the Result type. I wonder if a more explicit name would be `try_usize_to_i32` to indicate the `Result` return type, but that's just thinking out loud for now.

Reply:
I like that a lot, actually. It fits with the naming conventions from the TryFrom trait, making it signify its intent exactly.